### PR TITLE
Fix MTEBEvaluator: device mapping, padding-free inference, last-token pooling, L2 normalization

### PR DESCRIPTION
**Pull request by natke**
Fixes several issues in the MTEBEvaluator for embedding model evaluation:

## Device mapping
Maps Olive's `Device.GPU` (`"gpu"`) to PyTorch's `"cuda"` when initializing `SentenceTransformer` in the HF evaluation path. Also handles indexed devices (e.g. `gpu:0` → `cuda:0`).

## Padding-free inference for GenAI
GenAI's `Generator` does not accept an `attention_mask`, so padded batches produce contaminated hidden states via self-attention to padding tokens. Fix: process each sentence individually with only its real tokens, eliminating padding entirely.

## Last-token pooling
Replaced mean pooling with last-token pooling in the GenAI and ORT wrappers to match models like Qwen3-Embedding that use `pooling_mode_lasttoken=True`.

## L2 normalization
Added L2 normalization after pooling in the base `encode()` method, matching the `2_Normalize` module in the SentenceTransformer pipeline.

### Results
These fixes close the score gap between HF and GenAI evaluation:
- **Before**: HF 0.785 vs GenAI 0.651 (STS17 main_score)
- **After**: HF 0.785 vs GenAI 0.785

### CODE REVIEW

File: olive/evaluator/mteb_ort.py [hunk 1]
```diff
@@ -181,16 +185,15 @@ def _encode_batch(self, encoded_input: dict) -> np.ndarray:
...
     @staticmethod
-    def _mean_pool(hidden_states: np.ndarray, attention_mask: np.ndarray) -> np.ndarray:
-        """Mean pooling: average token embeddings weighted by attention mask."""
-        mask_expanded = np.expand_dims(attention_mask, axis=-1).astype(hidden_states.dtype)
-        sum_embeddings = np.sum(hidden_states * mask_expanded, axis=1)
-        sum_mask = np.clip(mask_expanded.sum(axis=1), a_min=1e-9, a_max=None)
-        return sum_embeddings / sum_mask
+    def _last_token_pool(hidden_states: np.ndarray, attention_mask: np.ndarray) -> np.ndarray:
+        """Last-token pooling: take the hidden state at the last non-padding token position."""
+        sequence_lengths = attention_mask.sum(axis=1).astype(int) - 1
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/mteb_ort.py]: fixed

File: olive/evaluator/mteb_ort.py [hunk 2]
```diff
@@ -230,34 +233,30 @@ def __init__(
...
+        all_embeddings = []
+        for i in range(batch_size):
+            real_len = int(attention_mask[i].sum())
+            ids = input_ids[i, :real_len].reshape(1, -1)
+
+            params = og.GeneratorParams(self.model)
+            params.set_search_options(max_length=real_len + 1, past_present_share_buffer=False, batch_size=1)
+
+            generator = og.Generator(self.model, params)
+            generator.append_tokens(ids.tolist())
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/mteb_ort.py]: fixed

File: olive/evaluator/mteb_ort.py [hunk 3]
```diff
@@ -84,7 +84,11 @@ def encode(self, inputs, *, task_metadata=None, hf_split=None, hf_subset=None, p
             embeddings = self._encode_batch(encoded)
             all_embeddings.append(embeddings)
 
-        return np.concatenate(all_embeddings, axis=0)
+        result = np.concatenate(all_embeddings, axis=0)
+        # L2 normalize (matches SentenceTransformer Normalize module)
+        norms = np.linalg.norm(result, axis=1, keepdims=True)
+        norms = np.clip(norms, a_min=1e-9, a_max=None)
+        return result / norms
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/mteb_ort.py]: fixed

File: olive/evaluator/mteb_ort.py [hunk 4]
```diff
@@ -230,34 +241,48 @@
...
-        params.set_search_options(max_length=seq_len + 1, past_present_share_buffer=False, batch_size=batch_size)
 
-        generator = og.Generator(self.model, params)
-        generator.append_tokens(input_ids.tolist())
+        # GenAI Generator does not accept attention_mask, so padding tokens
+        # contaminate hidden states via self-attention. To avoid this, group
+        # sequences by real length and process each group as a single batch
+        # (no padding needed within a group of equal-length sequences).
+        real_lengths = attention_mask.sum(axis=1).astype(int)
+        embeddings = np.empty((batch_size, 0), dtype=np.float32)  # placeholder, filled below
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/mteb_ort.py]: fixed

[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/mteb_ort.py]: resolved

**Review by jambayk — APPROVED** [review]
